### PR TITLE
feat: integration dokka

### LINKED ISSUE #354: integration dokka
When integrating Dokka and attempting to execute ./gradlew dokkaHtmlMultiModule, there are some compilation errors.
WX20230719-111646
we need to change like this 
```kotlin
inline fun <K : Any, V : Any> LruCache<K, V>.getOrPut(key: K, defaultValue: () -> V): V {
    synchronized(this) {
        this[key]?.let { return it }
        return defaultValue().apply { put(key, this) }
    }
}

/**
 * Gets value with specific key from the cache. If the value is not present,
 * calls [defaultValue] to obtain a value which is placed into the cache
 * if not null, then returned.
 *
 * This method is thread-safe.
 */
inline fun <K : Any, V> LruCache<K, V>.getOrPutNotNull(key: K, defaultValue: () -> V?): V? {
    synchronized(this) {
        this[key]?.let { return it }
        return defaultValue()?.apply { put(key, this) }
    }
}
```
then it works great.

### PR DESCRIPTION
**Pull request by 5peak2me**
fix #354

### CODE REVIEW

**Comment by FunkyMuse** [comment]
does `dokkaHtmlMultiModule` succeeds?

Where do we publish the docs?

**Comment by 5peak2me** [comment]
Yeah, this PR has fixed the issue with `dokkaHtmlMultiModule` not running. The current doc is located in `build/dokka`. I suggest creating a repository called `KAHelpers.github.io`, where we can publish the documentation.

**Comment by FunkyMuse** [comment]
> Yeah, this PR has fixed the issue with `dokkaHtmlMultiModule` not running. The current doc is located in `build/dokka`. I suggest to enable pages feature, where we can publish the documentation.

Have any tutorial how to achieve this?

**Comment by 5peak2me** [comment]
> > Yeah, this PR has fixed the issue with `dokkaHtmlMultiModule` not running. The current doc is located in `build/dokka`. I suggest to enable pages feature, where we can publish the documentation.
> 
> Have any tutorial how to achieve this?

seeing the `Project site` tab in this: https://pages.github.com/

**Comment by FunkyMuse** [comment]
> > > Yeah, this PR has fixed the issue with `dokkaHtmlMultiModule` not running. The current doc is located in `build/dokka`. I suggest to enable pages feature, where we can publish the documentation.
> > 
> > 
> > Have any tutorial how to achieve this?
> 
> seeing the `Project site` tab in this: https://pages.github.com/

then you need to configure the plugin to create the documentation outside the `/build/dokka` folder since `/build` is omitted from the Git tree, also I remember back then that Dokka had a terrible CSS and it looked ugly, don't know what's the case now.

Also the plugin must be applied for ` case ["enums", "regex"]:` subprojects too

**Comment by 5peak2me** [comment]
I've created a [sample](https://github.com/5peak2me/dokka-github-pages-workflows) repository and outlined the specific implementation steps. I'll be submitting a pull request (PR) with the changes. The documentation style can be found here [dokka-github-pages-workflows](https://5peak2me.github.io/dokka-github-pages-workflows/)

File: build.gradle [hunk 1]
```diff
@@ -66,7 +66,9 @@ buildscript {
     }
 }
 
-
+plugins {
+    id 'org.jetbrains.dokka' version '1.8.20'
```
Comment (FunkyMuse) [inline build.gradle:70]: For the version you can use `kotlin_version`

Comment (5peak2me) [reply build.gradle:70]: According to [Dokka](https://github.com/Kotlin/dokka), the stable release is `1.8.20`. Maybe later, we can change the dependencies system to version catlaogs.

**Comment by 5peak2me** [comment]
Release a tag called 3.4.9. it works.

WX20230721-094323
WX20230721-094343

**Comment by FunkyMuse** [comment]
> Release a tag called 3.4.9. it works.
> 
> WX20230721-094323 WX20230721-094343

great, now we can have it changed to `main` only instead of tags and also the plugin must be applied for  case ["enums", "regex"]: subprojects too

**Comment by FunkyMuse** [comment]
> I didn't get the meaning of `the plugin must be applied for case ["enums", "regex"]: subprojects`, can you explain?

There are Kotlin/Java only modules, we should have docs for them too.

**Comment by FunkyMuse** [comment]
> > > I didn't get the meaning of `the plugin must be applied for case ["enums", "regex"]: subprojects`, can you explain?
> > 
> > 
> > There are Kotlin/Java only modules, we should have docs for them too.
> 
> I get it, I'll test it later. It might take some time. Or we can run dokka first and if it works, I can submit another PR.

It's one line change, just add `apply plugin: 'org.jetbrains.dokka'`

And also change the pipeline to run on main push

**Comment by 5peak2me** [comment]
done

File: .github/workflows/documentation.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+name: Documentation
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
```
Comment (FunkyMuse) [inline .github/workflows/documentation.yml:7]: Not needed

**Review by FunkyMuse — CHANGES_REQUESTED** [review]

File: .github/workflows/documentation.yml (hunk 2 above)
Comment (5peak2me) [reply .github/workflows/documentation.yml]: done

**Pull request merged**